internal/controllers/machineset/machineset_controller.go

```diff
@@ -674,145 +834,330 @@ func (r *Reconciler) syncReplicas(ctx context.Context, s *scope) (ctrl.Result, e
+					continue
+				}
+				notAcknowledgeMoveReplicas.Insert(m.Name)
 			}
+		}
+		if notAcknowledgeMoveReplicas.Len() > 0 {
+			log.Info(fmt.Sprintf("Replicas %s moved from an old MachineSet still pending acknowledge from machine deployment", notAcknowledgeMoveReplicas.UnsortedList()))
+		}
 
-			s.scaleUpPreflightCheckErrMessages = preflightCheckErrMessages
-			v1beta1conditions.MarkFalse(ms, clusterv1.MachinesCreatedV1Beta1Condition, clusterv1.PreflightCheckFailedV1Beta1Reason, clusterv1.ConditionSeverityError, "%s", strings.Join(preflightCheckErrMessages, "; "))
-			if err != nil {
-				return ctrl.Result{}, err
-			}
-			return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+		machinesToDeleteOrMove := len(machines) - notAcknowledgeMoveReplicas.Len() - int(ptr.Deref(ms.Spec.Replicas, 0))
+		if machinesToDeleteOrMove == 0 {
+			return ctrl.Result{}, nil
+		}
+
+		// Move machines to the target MachineSet if the current MachineSet is instructed to do so.
+		if targetMSName, ok := ms.Annotations[clusterv1.MachineSetMoveMachinesToMachineSetAnnotation]; ok && targetMSName != "" {
+			// Note: The number of machines actually moved could be less than expected e.g. because some machine still updating in-place from a previous move.
+			return r.startMoveMachines(ctx, s, targetMSName, machinesToDeleteOrMove)
+		}
+
+		// Otherwise the current MachineSet is not instructed to move machines to another MachineSet,
+		// then delete all the exceeding machines.
+		return r.deleteMachines(ctx, s, machinesToDeleteOrMove)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) createMachines(ctx context.Context, s *scope, machinesToAdd int) (ctrl.Result, error) {
+	if r.overrideCreateMachines != nil {
+		return r.overrideCreateMachines(ctx, s, machinesToAdd)
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	ms := s.machineSet
+	cluster := s.cluster
+
+	preflightCheckErrMessages, err := r.runPreflightChecks(ctx, cluster, s.machineSet, "Scale up")
+	if err != nil || len(preflightCheckErrMessages) > 0 {
+		if err != nil {
+			// If err is not nil use that as the preflightCheckErrMessage
+			preflightCheckErrMessages = append(preflightCheckErrMessages, err.Error())
+		}
+
+		s.scaleUpPreflightCheckErrMessages = preflightCheckErrMessages
+		v1beta1conditions.MarkFalse(ms, clusterv1.MachinesCreatedV1Beta1Condition, clusterv1.PreflightCheckFailedV1Beta1Reason, clusterv1.ConditionSeverityError, "%s", strings.Join(preflightCheckErrMessages, "; "))
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	machinesAdded := []*clusterv1.Machine{}
+	for i := range machinesToAdd {
+		// Create a new logger so the global logger is not modified.
+		log := log
+		machine, computeMachineErr := r.computeDesiredMachine(ms, nil)
+		if computeMachineErr != nil {
+			r.recorder.Eventf(ms, corev1.EventTypeWarning, "FailedCreate", "Failed to create machine: %v", err)
```

Comment:
If computeDesiredMachine fails it will probably fail pretty consistently and very often. Is it okay to produce events with that frequency?

Similar question for other events added in this PR (except the ones that occur only rarely)

Reply:
dropped